pydis_site/apps/api/serializers.py

```diff
@@ -249,15 +255,57 @@ class Meta:
         fields = ('id', 'name', 'colour', 'permissions', 'position')
 
 
-class UserSerializer(BulkSerializerMixin, ModelSerializer):
+class UserListSerializer(ListSerializer):
+    """List serializer for User model to handle bulk updates."""
+
+    def create(self, validated_data: list) -> list:
+        """Override create method to optimize django queries."""
+        present_users = User.objects.all()
+        new_users = []
+        present_user_ids = [user.id for user in present_users]
+
+        for user_dict in validated_data:
+            if user_dict["id"] in present_user_ids:
+                raise ValidationError({"id": "User already exists."})
+            new_users.append(User(**user_dict))
+
+        return User.objects.bulk_create(new_users)
+
+    def update(self, instance: QuerySet, validated_data: list) -> list:
+        """
+        Override update method to support bulk updates.
+
+        ref:https://www.django-rest-framework.org/api-guide/serializers/#customizing-multiple-update
+        """
+        instance_mapping = {user.id: user for user in instance}
+        data_mapping = {item['id']: item for item in validated_data}
+
+        updated = []
+        fields_to_update = set()
+        for user_id, data in data_mapping.items():
+            for key in data:
+                fields_to_update.add(key)
+            user = instance_mapping.get(user_id)
+            user.__dict__.update(data)
+            updated.append(user)
+
+        fields_to_update.remove("id")
+        User.objects.bulk_update(updated, fields_to_update)
```

Comment:
I believe this should catch `IntegrityError` and raise a `ValidatioError` since DRF doesn't handle it by default. This may be something that can be handled at a more global scale with a custom error handler (which I think is ideal) if you want to look into it further.

Reply:
I found that, if no fields are provided, a validation error is thrown.
How does the IntegrityError get triggered?

Reply:
When a DB constraint is violated.